File: apps/mobile/modules/kilo-launcher-surfaces/ios/KiloLauncherSurfacesModule.swift
import ExpoModulesCore
import UIKit

/**
 * The iOS half of the launcher surfaces: the Home Screen Quick Actions a long
 * press on the app icon shows.
 *
 * JS owns the urls, the translated labels and the decision of when an agent
 * waits; this module owns what the launcher shows. The shortcut `type` is a
 * stable per-action identifier and the deep-link url rides in `userInfo`, so
 * native holds no route mapping and a tap re-enters the app through the same
 * pipeline a notification tap uses. An action whose url is null is absent, which
 * is how "Needs input" disappears once nothing waits.
 *
 * iOS has no quick-settings tile; its twin is a Control Center control, which is
 * a separate request and deliberately not built here.
 */
public final class KiloLauncherSurfacesModule: Module {
  public func definition() -> ModuleDefinition {
    Name("KiloLauncherSurfaces")

    // Kept synchronous to match the fire-and-forget contract JS declares for
    // both of these (src/lib/native-launcher-surfaces.ts). `UIApplication` is
    // main-thread-only and expo-modules-core 57 offers `runOnQueue` on async
    // functions only, so the UIKit write hops to the main queue rather than
    // changing the JS-visible shape.
    Function("setSurfaces") { (payloadJson: String) in
      self.publish(payloadJson)
    }

    Function("clearDynamicSurfaces") {
      self.clearDynamicSurfaces()
    }

    // The cold-start handoff. Android has no such function: its shortcut and
    // tile intents arrive as deep links instead.
    Function("consumePendingLaunchUrl") { () -> String? in
      KiloLauncherSurfacesStore.consumePendingLaunchUrl()
    }
  }

  private func publish(_ payloadJson: String) {
    guard let surfaces = KiloLauncherSurfacesPayload(json: payloadJson) else {
      return
    }
    KiloLauncherSurfacesStore.cachePayload(payloadJson)
    setShortcutItems(surfaces.shortcutItems)
  }

  /** Called by JS after sign-out: keep only the cached New agent action. */
  private func clearDynamicSurfaces() {
    // The parked cold-start url belongs to the account that is leaving: its only
    // reader, `consumePendingLaunchUrl`, runs on the signed-in mount, so a url
    // left behind would route the next account to the previous account's
    // session. Drop it with the dynamic shortcuts it came from.
    KiloLauncherSurfacesStore.clearPendingLaunchUrl()
    let cached = KiloLauncherSurfacesStore.cachedPayloadJson().flatMap {
      KiloLauncherSurfacesPayload(json: $0)
    }
    setShortcutItems(cached?.newAgentItem.map { [$0] } ?? [])
  }

  private func setShortcutItems(_ items: [UIApplicationShortcutItem]) {
    DispatchQueue.main.async {
      UIApplication.shared.shortcutItems = items
    }
  }
}

/**
 * The Quick Action contract shared by the module and the app delegate subscriber.
 *
 * `type` cannot be the deep-link url: iOS requires a shortcut's type to be unique
 * across the app, and a session that is both the longest-waiting and the
 * last-opened one gives Needs input and Open last session the same url, so two of
 * the three actions would carry one type. The type is therefore a stable
 * per-action identifier and the url travels in `userInfo`.
 */
enum KiloLauncherSurfacesShortcut {
  static let newAgentType = "kilo.launcher.new-agent"
  static let needsInputType = "kilo.launcher.needs-input"
  static let openLastSessionType = "kilo.launcher.open-last-session"
  /** The `userInfo` key holding the deep-link url a Quick Action must open. */
  static let urlUserInfoKey = "kilo.launcher.url"
}

/**
 * The launcher surfaces payload frozen by the JS side. `newAgentUrl` and
 * `newAgentLabel` are always present; the other two actions are dynamic, and a
 * null or empty url means the action is absent and must not be offered.
 */
private struct KiloLauncherSurfacesPayload {
  let newAgentUrl: String?
  let newAgentLabel: String?
  let needsInputUrl: String?
  let needsInputLabel: String?
  let openLastSessionUrl: String?
  let openLastSessionLabel: String?

  init?(json: String) {
    guard let data = json.data(using: .utf8),
          let parsed = try? JSONSerialization.jsonObject(with: data),
          let object = parsed as? [String: Any] else {
      return nil
    }
    newAgentUrl = payloadString(object, "newAgentUrl")
    newAgentLabel = payloadString(object, "newAgentLabel")
    needsInputUrl = payloadString(object, "needsInputUrl")
    needsInputLabel = payloadString(object, "needsInputLabel")
    openLastSessionUrl = payloadString(object, "openLastSessionUrl")
    openLastSessionLabel = payloadString(object, "openLastSessionLabel")
  }

  /** At most three items, in publish order. A null url omits its action. */
  var shortcutItems: [UIApplicationShortcutItem] {
    [newAgentItem, needsInputItem, openLastSessionItem].compactMap { $0 }
  }

  var newAgentItem: UIApplicationShortcutItem? {
    shortcut(
      identifier: KiloLauncherSurfacesShortcut.newAgentType,
      url: newAgentUrl,
      label: newAgentLabel
    )
  }

  private var needsInputItem: UIApplicationShortcutItem? {
    shortcut(
      identifier: KiloLauncherSurfacesShortcut.needsInputType,
      url: needsInputUrl,
      label: needsInputLabel
    )
  }

  private var openLastSessionItem: UIApplicationShortcutItem? {
    shortcut(
      identifier: KiloLauncherSurfacesShortcut.openLastSessionType,
      url: openLastSessionUrl,
      label: openLastSessionLabel
    )
  }

  /**
   * The title is the label JS translated; the url is only the fallback for a
   * missing label, never an English literal. The type is the action identifier,
   * not the url, so two actions opening the same session keep distinct types.
   */
  private func shortcut(
    identifier: String,
    url: String?,
    label: String?
  ) -> UIApplicationShortcutItem? {
    guard let url else {
      return nil
    }
    return UIApplicationShortcutItem(
      type: identifier,
      localizedTitle: label ?? url,
      localizedSubtitle: nil,
      icon: nil,
      // The url rides in `userInfo` as the NSString the `NSSecureCoding` value
      // type accepts, so two actions opening the same session keep distinct types.
      userInfo: [KiloLauncherSurfacesShortcut.urlUserInfoKey: url as NSString]
    )
  }
}

/** Reads a payload string, treating an explicit JSON null or an empty string as absent. */
private func payloadString(_ object: [String: Any], _ key: String) -> String? {
  guard let value = object[key] as? String, !value.isEmpty else {
    return nil
  }
  return value
}

File: apps/mobile/modules/kilo-launcher-surfaces/android/src/main/java/com/kilocode/kilaunchersurfaces/KiloLauncherSurfacesModule.kt
package com.kilocode.kilaunchersurfaces

import android.content.ComponentName
import android.content.Context
import android.content.pm.ShortcutInfo
import android.content.pm.ShortcutManager
import android.graphics.drawable.Icon
import android.os.Build
import android.service.quicksettings.TileService
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import org.json.JSONObject

/**
 * Local Expo module for the Android launcher surfaces: the dynamic app shortcuts
 * on the icon and the quick-settings tile.
 *
 * JS owns the URLs, the translated labels and the decision of when an agent waits;
 * this module owns the native caches and pins the launcher surfaces. The payload is
 * committed synchronously because [QuickSettingsTileService] reads it in a cold
 * process, and the shortcuts are published with `setDynamicShortcuts` so an action
 * whose url is null disappears from the launcher.
 *
 * App shortcuts are a framework API from API 25, so an API 24 device shows none;
 * the tile (API 24+) still works there.
 */
class KiloLauncherSurfacesModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("KiloLauncherSurfaces")

    Function("setSurfaces") { payloadJson: String ->
      cacheSurfaces(payloadJson)
    }

    Function("clearDynamicSurfaces") {
      clearDynamicSurfaces()
    }
  }

  // `AppContext` exposes only the React context. Every entry point runs from a JS
  // call, so losing it means the module cannot work at all.
  private val context: Context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

  private fun cacheSurfaces(payloadJson: String) {
    val payload = JSONObject(payloadJson)
    val surfaces =
      CachedSurfaces(
        newAgentUrl = payload.stringOrNull(LauncherSurfacesStore.KEY_NEW_AGENT_URL),
        newAgentLabel = payload.stringOrNull(LauncherSurfacesStore.KEY_NEW_AGENT_LABEL),
        needsInputUrl = payload.stringOrNull(LauncherSurfacesStore.KEY_NEEDS_INPUT_URL),
        needsInputLabel = payload.stringOrNull(LauncherSurfacesStore.KEY_NEEDS_INPUT_LABEL),
        openLastSessionUrl = payload.stringOrNull(LauncherSurfacesStore.KEY_OPEN_LAST_SESSION_URL),
        openLastSessionLabel = payload.stringOrNull(LauncherSurfacesStore.KEY_OPEN_LAST_SESSION_LABEL),
      )

    val editor =
      LauncherSurfacesStore.prefs(context)
        .edit()
        .putOrRemove(LauncherSurfacesStore.KEY_NEW_AGENT_URL, surfaces.newAgentUrl)
        .putOrRemove(LauncherSurfacesStore.KEY_NEW_AGENT_LABEL, surfaces.newAgentLabel)
        .putOrRemove(LauncherSurfacesStore.KEY_NEEDS_INPUT_URL, surfaces.needsInputUrl)
        .putOrRemove(LauncherSurfacesStore.KEY_NEEDS_INPUT_LABEL, surfaces.needsInputLabel)
        .putOrRemove(LauncherSurfacesStore.KEY_OPEN_LAST_SESSION_URL, surfaces.openLastSessionUrl)
        .putOrRemove(LauncherSurfacesStore.KEY_OPEN_LAST_SESSION_LABEL, surfaces.openLastSessionLabel)
    check(editor.commit()) { "Cannot cache the launcher surfaces payload" }

    pinDynamicShortcuts(surfaces)
    requestTileRefresh()
  }

  /** Called by JS after sign-out: keep only the New agent action. */
  private fun clearDynamicSurfaces() {
    val editor =
      LauncherSurfacesStore.prefs(context)
        .edit()
        .remove(LauncherSurfacesStore.KEY_NEEDS_INPUT_URL)
        .remove(LauncherSurfacesStore.KEY_NEEDS_INPUT_LABEL)
        .remove(LauncherSurfacesStore.KEY_OPEN_LAST_SESSION_URL)
        .remove(LauncherSurfacesStore.KEY_OPEN_LAST_SESSION_LABEL)
    check(editor.commit()) { "Cannot clear the launcher surfaces payload" }

    pinDynamicShortcuts(LauncherSurfacesStore.read(context))
    requestTileRefresh()
  }

  private fun pinDynamicShortcuts(surfaces: CachedSurfaces) {
    if (Build.VERSION.SDK_INT < 25) {
      return
    }
    val shortcuts = mutableListOf<ShortcutInfo>()
    surfaces.newAgentUrl?.let {
      shortcuts += shortcut(SHORTCUT_NEW_AGENT, surfaces.newAgentLabel, it)
    }
    surfaces.needsInputUrl?.let {
      shortcuts += shortcut(SHORTCUT_NEEDS_INPUT, surfaces.needsInputLabel, it)
    }
    surfaces.openLastSessionUrl?.let {
      shortcuts += shortcut(SHORTCUT_OPEN_LAST_SESSION, surfaces.openLastSessionLabel, it)
    }
    shortcutManager().setDynamicShortcuts(shortcuts)
  }

  private fun shortcut(id: String, label: String?, url: String): ShortcutInfo {
    val builder =
      ShortcutInfo.Builder(context, id)
        .setShortLabel(label ?: id)
        .setLongLabel(label ?: id)
        .setIcon(Icon.createWithResource(context, R.drawable.kilo_launcher_surfaces_shortcut))
        .setIntent(actionIntent(context, url))
    // setLongLived(boolean) is API 29: unguarded it throws NoSuchMethodError on an
    // API 25-28 device and the whole shortcut publish (and the tile refresh) is lost.
    if (Build.VERSION.SDK_INT >= 29) {
      builder.setLongLived(true)
    }
    return builder.build()
  }

  private fun shortcutManager(): ShortcutManager =
    context.getSystemService(Context.SHORTCUT_SERVICE) as ShortcutManager

  private fun requestTileRefresh() {
    TileService.requestListeningState(
      context,
      ComponentName(context, QuickSettingsTileService::class.java),
    )
  }

  private companion object {
    const val SHORTCUT_NEW_AGENT = "new-agent"
    const val SHORTCUT_NEEDS_INPUT = "needs-input"
    const val SHORTCUT_OPEN_LAST_SESSION = "open-last-session"
  }
}
